xds/src/main/java/io/grpc/xds/sds/TlsContextManager.java

```diff
@@ -45,12 +63,42 @@ public static synchronized TlsContextManager getInstance() {
   /** Creates a SslContextProvider. Used for retrieving a server-side SslContext. */
   public SslContextProvider findOrCreateServerSslContextProvider(
       DownstreamTlsContext downstreamTlsContext) {
-    return SecretVolumeSslContextProvider.getProviderForServer(downstreamTlsContext);
+    checkNotNull(downstreamTlsContext, "downstreamTlsContext");
+    return mapForServers.get(downstreamTlsContext);
   }
 
   /** Creates a SslContextProvider. Used for retrieving a client-side SslContext. */
   public SslContextProvider findOrCreateClientSslContextProvider(
       UpstreamTlsContext upstreamTlsContext) {
-    return SecretVolumeSslContextProvider.getProviderForClient(upstreamTlsContext);
+    checkNotNull(upstreamTlsContext, "upstreamTlsContext");
+    return mapForClients.get(upstreamTlsContext);
+  }
+
+  /** Releases an instance of the given client-side {@link SslContextProvider}.
+   *
+   *  <p>The instance must have been obtained from {@link #findOrCreateClientSslContextProvider}.
+   *  Otherwise will throw IllegalArgumentException.
+   *
+   *  <p>Caller must not release a reference more than once. It's advised that you clear the
+   *  reference to the instance with the null returned by this method.
+   */
+  public SslContextProvider releaseClientSslContextProvider(SslContextProvider sslContextProvider) {
+    checkNotNull(sslContextProvider, "sslContextProvider");
+    return mapForClients.release(
+        (UpstreamTlsContext) sslContextProvider.getKey(), sslContextProvider);
+  }
+
+  /** Releases an instance of the given server-side {@link SslContextProvider}.
+   *
+   *  <p>The instance must have been obtained from {@link #findOrCreateServerSslContextProvider}.
+   *  Otherwise will throw IllegalArgumentException.
+   *
+   *  <p>Caller must not release a reference more than once. It's advised that you clear the
+   *  reference to the instance with the null returned by this method.
+   */
+  public SslContextProvider releaseServerSslContextProvider(SslContextProvider sslContextProvider) {
+    checkNotNull(sslContextProvider, "sslContextProvider");
+    return mapForServers.release(
```

Comment:
for this design, we can make the map takes sslContextProvider which can get the Key. although this seems a bit after thought, BiMap could be useful here but less depends on guava is better. any idea improve this api? Also, if we can avoid object type to ensure type safety, it would be really nice.

Reply:
I did give it a lot of thought and this was the best (according to me) of all the alternatives.I wasn't happy about the `Object` to `DownstreamTlsContext` cast either.

The RefCountingMap actually verifies the key+value (SslContextProvider) passed by the caller are consistent so instead of that check we just have the Map get the key from the value and perform release - I am okay with that. But the cast will still remain (from Object to K which is the key-type in the Map).

Reply:
moved getKey into RefCountingMap.release

Reply:
if `SslContextProvider` accepts a generic type, it can be solved. I am not sure it is the best way to fix the type safety issue. it is probably better than type unsafe code. also, the field name `key` is very `ReferenceCountingSslContextProviderMap` oriented name.

Reply:
Regarding `SslContextProvider` accepting a generic type, that doesn't work as well because of type erasure. I will have to subclass `SslContextProvider` just based on the type of the `key` that's a member variable. So something like:

```
class SslContextProvider<K> {
K key;
}

class SslContextProviderForServer extends SslContextProvider<DownstreamTlsContext> {
}

class SslContextProviderForClient extends SslContextProvider<UpstreamTlsContext> {
}
```

That will provide the type-awareness for `key` that we are looking for. I think that's an overkill.

I agree with naming of `key` and did consider it.

Reply:
`ReferenceCountingSslContextProviderMap` is already taking generic K, so the release just need to be `public SslContextProvider release(final SslContextProvider<K> value)`. during the compile time the type will be checked. maybe i am misunderstanding or missing something?

Reply:
That might work. In my thinking `SslContextProvider` is not a concrete class but the subclass `SecretVolumeSslContextProvider` is and that will also need to be made generic with the `K` being propagated only to the superclass `SslContextProvider`. So this will be:
```
final class SecretVolumeSslContextProvider<K> extends SslContextProvider<K> {
}
```
And I can then just use `SecretVolumeSslContextProvider<DownstreamTlsContext>` vs `SecretVolumeSslContextProvider<UpstreamTlsContext>`

Okay, let me try that!

Reply:
fixed.